core-java/src/main/java/eu/neverblink/jelly/core/internal/NameDecoderImpl.java

```diff
@@ -0,0 +1,157 @@
+package eu.neverblink.jelly.core.internal;
+
+import eu.neverblink.jelly.core.NameDecoder;
+import eu.neverblink.jelly.core.RdfProtoDeserializationError;
+import eu.neverblink.jelly.core.proto.v1.RdfNameEntry;
+import eu.neverblink.jelly.core.proto.v1.RdfPrefixEntry;
+import java.util.function.Function;
+
+/**
+ * Class for decoding RDF IRIs from their Jelly representation.
+ *
+ * @param <TIri> The type of the IRI in the target RDF library.
+ */
+final class NameDecoderImpl<TIri> implements NameDecoder<TIri> {
+
+    private static final class NameLookupEntry {
+
+        // Primary: the actual name
+        public String name;
+        // Secondary values (may be mutated without invalidating the primary value)
+        // Reference to the last prefix ID used to encode the IRI with this name
+        public int lastPrefixId;
+        // Serial number of the last prefix ID used to encode the IRI with this name
+        public int lastPrefixSerial;
+        // Last IRI encoded with this name
+        public Object lastIri;
+    }
+
+    private static final class PrefixLookupEntry {
+
+        public String prefix;
+        public int serial = -1;
+    }
+
+    private final NameLookupEntry[] nameLookup;
+    private final PrefixLookupEntry[] prefixLookup;
+
+    private int lastPrefixIdReference = 0;
+    private int lastNameIdReference = 0;
+
+    private int lastPrefixIdSet = 0;
+    private int lastNameIdSet = 0;
+
+    private final Function<String, TIri> iriFactory;
+
+    /**
+     * Creates a new NameDecoder.
+     *
+     * @param prefixTableSize The size of the prefix lookup table.
+     * @param nameTableSize   The size of the name lookup table.
+     * @param iriFactory      A function that creates an IRI from a string.
+     */
+    public NameDecoderImpl(int prefixTableSize, int nameTableSize, Function<String, TIri> iriFactory) {
+        this.iriFactory = iriFactory;
+        nameLookup = new NameLookupEntry[nameTableSize + 1];
+        prefixLookup = new PrefixLookupEntry[prefixTableSize + 1];
+
+        for (int i = 1; i < nameTableSize + 1; i++) {
+            nameLookup[i] = new NameLookupEntry();
+        }
+        for (int i = 1; i < prefixTableSize + 1; i++) {
+            prefixLookup[i] = new PrefixLookupEntry();
+        }
+    }
+
+    /**
+     * Update the name table with a new entry.
+     *
+     * @param nameEntry name row
+     * @throws ArrayIndexOutOfBoundsException if the identifier is out of bounds
+     */
+    @Override
+    public void updateNames(RdfNameEntry nameEntry) {
```

Comment:
!!!!!! I'm not seeing the changes from #339 https://github.com/Jelly-RDF/jelly-jvm/commit/d90f502d31561d0a693fee1e8a899b6d33df214c

Please make sure you are including EVERYTHING from main as it is now, as there were some changes since you started implementing this.